airbyte-integrations/connectors/source-gitlab/source_gitlab/streams.py

```diff
@@ -0,0 +1,334 @@
+        return record
+
+
+class IncrementalGitlabChildStream(GitlabChildStream):
+    state_checkpoint_interval = 100
+    cursor_field = "updated_at"
+    filter_field = "updated_after"
+
+    def __init__(self, start_date, **kwargs):
+        super().__init__(**kwargs)
+        self._start_date = start_date
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        """
+        Return the latest state by comparing the cursor value in the latest record with the stream's most recent state object
+        and returning an updated state object.
+        """
+        project_id = latest_record.get("project_id")
+        latest_cursor_value = latest_record.get(self.cursor_field)
+        current_state = current_stream_state.get(str(project_id))
+        if current_state:
+            current_state = current_state.get(self.cursor_field)
+        current_state_value = current_state or latest_cursor_value
+        current_stream_state[str(project_id)] = {self.cursor_field: max(current_state_value, latest_cursor_value)}
+        return current_stream_state
+
+    def request_params(self, stream_state=None, stream_slice: Mapping[str, Any] = None, **kwargs):
+        stream_state = stream_state or {}
+        params = super().request_params(stream_state, stream_slice, **kwargs)
+
+        start_point = self._start_date
+        state_project_value = stream_state.get(str(stream_slice["id"]))
+        if state_project_value:
+            state_value = state_project_value.get(self.cursor_field)
+            if state_value:
+                start_point = max(start_point, state_value)
+        params[self.filter_field] = start_point
+        return params
+
+
+class Groups(GitlabStream):
+    def __init__(self, group_ids: List, **kwargs):
+        super().__init__(**kwargs)
+        self.group_ids = group_ids
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"groups/{stream_slice['id']}"
+
+    def stream_slices(self, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
+        for gid in self.group_ids:
+            yield {"id": gid}
+
+    def transform(self, record, stream_slice: Mapping[str, Any] = None, **kwargs):
+        record["projects"] = [
+            {"id": project["id"], "path_with_namespace": project["path_with_namespace"]} for project in record.pop("projects", [])
+        ]
+        return record
+
+
+class Projects(GitlabStream):
+    stream_base_params = {"statistics": 1}
+
+    def __init__(self, project_ids: List = None, parent_stream: GitlabStream = None, **kwargs):
+        super().__init__(**kwargs)
+        self.project_ids = project_ids
+        self.parent_stream = parent_stream
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return f"projects/{stream_slice['id']}"
+
+    def stream_slices(self, **kwargs) -> Iterable[Optional[Mapping[str, any]]]:
```

Comment:
this is very complicated, can't we have GroupProjects class, inherit it from Projects and move first part of IF there?